src/mame/drivers/alphatpx.cpp

```diff
@@ -453,33 +789,132 @@ uint32_t alphatpx_state::screen_update(screen_device &screen, bitmap_rgb32 &bitm
 	return 0;
 }
 
+//**************************************************************************
+//  SOUND - Alphatronic P1, P2, P2S, P2U and Hell 2069
+//**************************************************************************
+
+WRITE8_MEMBER( alphatp_12_state::beep_w )
+{
+	m_beep->set_state(data&1);
+}
 
 //**************************************************************************
-//  SOUND
+//  SOUND - Alphatronic P3, P4, P30 and P40
 //**************************************************************************
 
-WRITE8_MEMBER( alphatpx_state::beep_w )
+WRITE8_MEMBER( alphatp_34_state::beep_w )
 {
 	m_beep->set_state(data&1);
 }
 
 //**************************************************************************
-//  SOUND
+//  FLOPPY - Alphatronic P1, P2, P2S, P2U and Hell 2069
+//**************************************************************************
+
+WRITE_LINE_MEMBER(alphatp_12_state::fdcirq_w)
+{
+	m_fdc_irq = state;
+}
+
+#include "debugger.h"
+
+WRITE_LINE_MEMBER(alphatp_12_state::fdcdrq_w)
+{
+	m_fdc_drq = state;
+}
+
+WRITE_LINE_MEMBER(alphatp_12_state::fdchld_w)
+{
+	m_fdc_hld = state;
+}
+
+READ8_MEMBER(alphatp_12_state::fdc_stat_r)
+{
+	uint8_t res = 0;
+	floppy_image_device *floppy1,*floppy2,*floppy3;
+	floppy1 = floppy2 = floppy3 = nullptr;
+	
+	floppy1 = m_con1 ? m_con1->get_device() : nullptr;
+	floppy2 = m_con2 ? m_con2->get_device() : nullptr;
+	floppy3 = m_con3 ? m_con3->get_device() : nullptr;
+
+	res = m_fdc_drq ? 0x80 : 0x00;
+	res |= m_fdc_irq ? 0x40 : 0x00;
+	res |= m_fdc_hld ? 0x00 : 0x20;
+	if (floppy3) res |= !floppy3->ready_r() ? 0x10 : 0;
+	if (floppy2) res |= !floppy2->ready_r() ? 0x08 : 0;
+	if (floppy1) res |= !floppy1->ready_r() ? 0x04 : 0;
+	if (m_curfloppy) res |= m_curfloppy->wpt_r() ? 0x02 : 0;
+
+	return res;
+}
+
+/* As far as we can tell, the mess of ttl de-inverts the bus */
+READ8_MEMBER(alphatp_12_state::fdc_r)
+{
+	return m_fdc->gen_r(offset) ^ 0xff;
+}
+
+WRITE8_MEMBER(alphatp_12_state::fdc_w)
+{
+	m_fdc->gen_w(offset, data ^ 0xff);
+}
+
+
+WRITE8_MEMBER(alphatp_12_state::fdc_cmd_w)
+{
+	floppy_image_device *floppy = nullptr;
+
+	logerror("%02x to fdc_cmd_w: motor %d side %d\n", data, (data & 0x10)>>4, (data & 4)>>2);
+
+	// select drive
+	if (!(data & 0x80))
+		{
+		floppy = m_con1 ? m_con1->get_device() : nullptr;
+		}
+	else if (!(data & 0x40))
+		{
+		floppy = m_con2 ? m_con2->get_device() : nullptr;
+		}
+	else if (!(data & 0x20))
+		{
+		floppy = m_con3 ? m_con3->get_device() : nullptr;
+		}
```

Comment:
Please follow convention of indenting braces to the level of the control statement.